File: contact-form/server/src/validators/contactValidator.ts
import type { ValidationErrors } from '../types'

const MIN_NAME_LENGTH = 2
const MAX_NAME_LENGTH = 100
const MIN_MESSAGE_LENGTH = 10
const MAX_MESSAGE_LENGTH = 1000

const EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

export interface ValidationResult {
    isValid: boolean
    errors: ValidationErrors
}

const validateName = (name: string): string | undefined => {
    const trimmed = name.trim()

    if (!trimmed) {
        return 'Имя обязательно для заполнения'
    }

    if (trimmed.length < MIN_NAME_LENGTH) {
        return 'Имя должно содержать минимум 2 символа'
    }

    if (trimmed.length > MAX_NAME_LENGTH) {
        return 'Имя не должно превышать 100 символов'
    }

    return undefined
}

const validateEmail = (email: string): string | undefined => {
    const trimmed = email.trim()

    if (!trimmed) {
        return 'Email обязателен для заполнения'
    }

    if (!EMAIL_REGEX.test(trimmed)) {
        return 'Некорректный формат email'
    }

    return undefined
}

const validateMessage = (message: string): string | undefined => {
    const trimmed = message.trim()

    if (!trimmed) {
        return 'Сообщение обязательно для заполнения'
    }

    if (trimmed.length < MIN_MESSAGE_LENGTH) {
        return 'Сообщение должно содержать минимум 10 символов'
    }

    if (trimmed.length > MAX_MESSAGE_LENGTH) {
        return 'Сообщение не должно превышать 1000 символов'
    }

    return undefined
}

export const validateContactForm = (
    name: string,
    email: string,
    message: string
): ValidationResult => {
    const errors: ValidationErrors = {}

    const nameError = validateName(name)
    const emailError = validateEmail(email)
    const messageError = validateMessage(message)

    if (nameError) errors.name = nameError
    if (emailError) errors.email = emailError
    if (messageError) errors.message = messageError

    return {
        isValid: Object.keys(errors).length === 0,
        errors,
    }
}

File: contact-form-aurora/server/src/contactvalidator.cpp
#include "contactvalidator.h"

ContactValidator::ContactValidator()
    : m_emailRegex(QStringLiteral("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$"))
{
}

ValidationResult ContactValidator::validate(const QString &name, 
                                           const QString &email, 
                                           const QString &message) const
{
    ValidationResult result;
    result.nameError = validateName(name);
    result.emailError = validateEmail(email);
    result.messageError = validateMessage(message);
    return result;
}

QString ContactValidator::validateName(const QString &name) const
{
    QString trimmed = name.trimmed();
    
    if (trimmed.isEmpty()) {
        return QStringLiteral("Имя обязательно для заполнения");
    }
    
    if (trimmed.length() < MIN_NAME_LENGTH) {
        return QStringLiteral("Имя должно содержать минимум 2 символа");
    }
    
    if (trimmed.length() > MAX_NAME_LENGTH) {
        return QStringLiteral("Имя не должно превышать 100 символов");
    }
    
    return QString();
}

QString ContactValidator::validateEmail(const QString &email) const
{
    QString trimmed = email.trimmed();
    
    if (trimmed.isEmpty()) {
        return QStringLiteral("Email обязателен для заполнения");
    }
    
    QRegularExpressionMatch match = m_emailRegex.match(trimmed);
    if (!match.hasMatch()) {
        return QStringLiteral("Некорректный формат email");
    }
    
    return QString();
}

QString ContactValidator::validateMessage(const QString &message) const
{
    QString trimmed = message.trimmed();
    
    if (trimmed.isEmpty()) {
        return QStringLiteral("Сообщение обязательно для заполнения");
    }
    
    if (trimmed.length() < MIN_MESSAGE_LENGTH) {
        return QStringLiteral("Сообщение должно содержать минимум 10 символов");
    }
    
    if (trimmed.length() > MAX_MESSAGE_LENGTH) {
        return QStringLiteral("Сообщение не должно превышать 1000 символов");
    }
    
    return QString();
}

